### DISCUSSION TITLE
Would it be faster to process entire corpus?

### DISCUSSION THREAD

**Discussion by the1gofer** [explosion/spaCy#7326] [Help: Best practices]
I have about 1600 documents that I'm wanting to process with Spacy, but it takes ages to generate the doc for that many strings.  Would it be faster to combine them all into one large text then slice them up somehow down the road?

**Comment by polm** [comment]
If you're running into speed issues with lots of documents we usually recommend using [nlp.pipe](https://spacy.io/api/language#pipe). From the docs:

```
texts = ["One document.", "...", "Lots of documents"]
for doc in nlp.pipe(texts, batch_size=50):
    assert doc.has_annotation("DEP")
```

How large are your individual documents?

**Reply by the1gofer** [reply]
That's a good question.  I don't know exactly.  Not very long.   A few paragraphs each.

**Reply by polm** [reply]
OK, that should be fine. Give `nlp.pipe` a try and let us know if it doesn't work.

**Reply by the1gofer** [reply]
is there any way to identify them as they are processed?    Let's say I load 100 documents, and then I want to compare one to every other document, and 2 to every other document, and so on.   When I get done I'll know that some of them are more similar than others, but it'll be difficult to know which ones specifically.   If I have id's for each one, can I tie them together some so when they come out the other end, I know document one and document fifteen are the same?

is that what the context in the (doc, context) is?  I don't see much about what it is specifically.

Edit:  that is what that does. Figured it out.

**Reply by polm** [reply]
I'm a little confused about what you're asking and what you figured out, but just in case, here's an answer:

To add IDs to docs, you can set [a custom attribute](https://spacy.io/usage/processing-pipelines#custom-components-attributes). That will let you track an individual doc through the pipeline.

"Context" usually refers to the tokens around another token, like the sentence that contains it, which is important for many kinds of predictions. Where did you see (doc, context)?

For finding most similar docs, unless you have a very small number of docs, you generally want to use some kind of approximate nearest neighbor search rather than doing a comparison with every other doc. [Annoy](https://spacy.io/usage/processing-pipelines#custom-components-attributes) is an easy to use library with that functionality. (100 is small enough that you probably don't need to use Annoy though.)

**Reply by the1gofer** [reply]
From [n.p.pipe](https://spacy.io/api/language#pipe):

- as_tuples | If set to True, inputs should be a sequence of (text, context) tuples. Output will then be a sequence of (doc, context) tuples. 

when I passed a (text, id), I got back (doc, id).   Unless I'm miss-understanding what's going on here.

**Reply by polm** [reply]
Ah, OK, I didn't realize that's where that came from. Yes, you've understood correctly, context there is just passing through whatever was in the input with the output.
